{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intro to object oriented programming\n",
    "### 1) what is it?\n",
    "Its a paradigm in software design to group data and functionality into objects\n",
    "that are easy to use and recycle and that provide the end user with access to complex functionality without having to worry about how it works.\n",
    "Typical objects that you deal with in coding are for example:\n",
    "- pandas data frames\n",
    "- matplotlib figures and axes\n",
    "- pathlib Path objects\n",
    "### 2) why you should know about it\n",
    "Unless you strat writing resuable code you wont need to worry about making yoru own classes. However ubderstanding the concept makes dealing with objects in your code a lot easier. For example you will undersatdn what is the difference between df.info() - a method, and df.columns - an attribute. Overall, it doesnt hurt looking a little bit more into how a computer programming language like python is designed if you want to become proficient in data science."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The python class\n",
    "At the heart of python OOP is the class. This tool allows you to design your\n",
    "own objects with teir own methods and attributes and undersatnd what you are doing when using\n",
    "other people's code.\n",
    "So for example when you say:\n",
    "df = pd.DataFrame\n",
    "of fig, ax = plt.subplots()\n",
    "you are dealing with classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.DnaSequence'>\n"
     ]
    }
   ],
   "source": [
    "#Lets make a simple class to deal with a DNA sequence for example:\n",
    "\n",
    "# The first set is to define the class:\n",
    "\n",
    "class DnaSequence: # Python classes are defined using names in CamelCase\n",
    "    pass\n",
    "\n",
    "# Now we can create an instance of the class:\n",
    "seq = DnaSequence()\n",
    "print(type(seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can lest do something with the class: We can have it hold data and also perform functions. So in our case for example, we would store metadata and nucleotide sequence and\n",
    "then perform steps such generating a reverse complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DnaSequence:\n",
    "    def __init__(self, sequence, species, gene_name):\n",
    "        self.sequence = sequence\n",
    "        self.length = len(sequence)\n",
    "        self.species = species\n",
    "        self.gene_name = gene_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok,, admittdely this looks strange. Whats the self? and the __init__? You dont have to worry too much about it. Self is just a way to refer to the object itself and attach things to it. the __init__ function is the way to define the basic attributes that are part of the class.\n",
    "Lets use the class now to instantite an actual sequence object that we can use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTTTGAACCCAGTTGGCGGGAGTGGCTGC\n",
      "30\n",
      "H. sapiens\n",
      "MASTL_mRNA\n"
     ]
    }
   ],
   "source": [
    "sequence = 'ACTTTGAACCCAGTTGGCGGGAGTGGCTGC'\n",
    "species = 'H. sapiens'\n",
    "gene_name = 'MASTL_mRNA'\n",
    "\n",
    "seq = DnaSequence(sequence, species, gene_name)\n",
    "print(seq.sequence)\n",
    "print(seq.length)\n",
    "print(seq.species)\n",
    "print(seq.gene_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So basically we have grouped a bunch of data (attributes) relelated to teh sequence together. Not very exciting! We could have used a simple dictionary to do this.\n",
    "Classes become more powerful when we give them functions (methods).\n",
    "So data connected to a class are called attributes, and functions methods.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTTTGAACCCAGTTGGCGGGAGTGGCTGC\n",
      "GCAGCCACTCCCGCCAACTGGGTTCAAAGT\n"
     ]
    }
   ],
   "source": [
    "class DnaSequence:\n",
    "    def __init__(self, sequence, species, gene_name):\n",
    "        self.sequence = sequence\n",
    "        self.length = len(sequence)\n",
    "        self.species = species\n",
    "        self.gene_name = gene_name\n",
    "\n",
    "    def reverse_complement(self):\n",
    "        complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        return ''.join([complement[base] for base in self.sequence[::-1]])\n",
    "    \n",
    "sequence = 'ACTTTGAACCCAGTTGGCGGGAGTGGCTGC'\n",
    "species = 'H. sapiens'\n",
    "gene_name = 'MASTL_mRNA'\n",
    "\n",
    "seq = DnaSequence(sequence, species, gene_name)\n",
    "print(seq.sequence) # note the attribute doen't need the parenthesis\n",
    "print(seq.reverse_complement()) # note the method needs the parenthesis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So whats the point of this?\n",
    "1. it can be useful to group attributes and functions together when writing more complex code. It makes it easuer to maintain and to refactor.\n",
    "2. Its the best way to generate functionsality that other people use, like pandas dataframes, or matplotlib figures and axes\n",
    "\n",
    "So now you should undersatnd that with df.columns you are accessing a class attribute and using df.info() or df.reset_index() ure are using a class method. So basically a function that someone wrote and connected with df to help you analyse data.\n",
    "\n",
    "Lets explore this a bit more by comparing a more complex bioinformatics program written as a step by step recipy in a notebook or as a class that we import from another file. \n",
    "Our aim now is to read in a Fasta file with sequence informationa and meta-data of a mRNA, and find the ORF in that sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NM_001172303.3 Homo sapiens microtubule associated serine/threonine kinase like (MASTL), transcript variant 1, mRNA\n",
      "ACTTTGAACCCAGTTGGCGGGAGTGGCTGCTCGCGGAGGGGCAGTGTCTGCGGGGCCGCTGTATGCTGTCCAGCGATGGATCCCACCGCGGGAAGCAAGAAGGAGCCTGGAGGAGGCGCGGCGACTGAGGAGGGCGTGAATAGGATCGCAGTGCCAAAACCGCCCTCCATTGAGGAATTCAGCATAGTGAAGCCCATTAGCCGGGGCGCCTTCGGGAAAGTGTATCTGGGGCAGAAAGGCGGCAAATTGTATGCAGTAAAGGTTGTTAAAAAAGCAGACATGATCAACAAAAATATGACTCATCAGGTCCAAGCTGAGAGAGATGCACTGGCACTAAGCAAAAGCCCATTCATTGTCCATTTGTATTATTCACTGCAGTCTGCAAACAATGTCTACTTGGTAATGGAATATCTTATTGGGGGAGATGTCAAGTCTCTCCTACATATATATGGTTATTTTGATGAAGAGATGGCTGTGAAATATATTTCTGAAGTAGCACTGGCTCTAGACTACCTTCACAGACATGGAATCATCCACAGGGACTTGAAACCGGACAATATGCTTATTTCTAATGAGGGTCATATTAAACTGACGGATTTTGGCCTTTCAAAAGTTACTTTGAATAGAGATATTAATATGATGGATATCCTTACAACACCATCAATGGCAAAACCTAGACAAGATTATTCAAGAACCCCAGGACAAGTGTTATCGCTTATCAGCTCGTTGGGATTTAACACACCAATTGCAGAAAAAAATCAAGACCCTGCAAACATCCTTTCAGCCTGTCTGTCTGAAACATCACAGCTTTCTCAAGGACTCGTATGCCCTATGTCTGTAGATCAAAAGGACACTACGCCTTATTCTAGCAAATTACTAAAATCATGTCTTGAAACAGTTGCCTCCAACCCAGGAATGCCTGTGAAGTGTCTAACTTCTAATTTACTCCAGTCTAGGAAAAGGCTGGCCACATCCAGTGCCAGTAGTCAATCCCACACCTTCATATCCAGTGTGGAATCAGAATGCCACAGCAGTCCCAAATGGGAAAAAGATTGCCAGGAAAGTGATGAAGCATTGGGCCCAACAATGATGAGTTGGAATGCAGTTGAAAAGTTATGCGCAAAATCTGCAAATGCCATTGAGACGAAAGGTTTCAATAAAAAGGATCTGGAGTTAGCTCTTTCTCCCATTCATAACAGCAGTGCCCTTCCCACCACTGGACGCTCTTGTGTAAACCTTGCTAAAAAATGCTTCTCTGGGGAAGTTTCTTGGGAAGCAGTAGAACTGGATGTAAATAATATAAATATGGACACTGACACAAGTCAGTTAGGTTTCCATCAGTCAAATCAGTGGGCTGTGGATTCTGGTGGGATATCTGAAGAGCACCTTGGGAAAAGAAGTTTAAAAAGAAATTTTGAGTTGGTTGACTCCAGTCCTTGTAAAAAAATTATACAGAATAAAAAAACTTGTGTAGAGTATAAGCATAACGAAATGACAAATTGTTATACAAATCAAAATACAGGCTTAACAGTTGAAGTGCAGGACCTTAAGCTATCAGTGCACAAAAGTCAACAAAATGACTGTGCTAATAAGGAGAACATTGTCAATTCTTTTACTGATAAACAACAAACACCAGAAAAATTACCTATACCAATGATAGCAAAAAACCTTATGTGTGAACTCGATGAAGACTGTGAAAAGAATAGTAAGAGGGACTACTTAAGTTCTAGTTTTCTATGTTCTGATGATGATAGAGCTTCTAAAAATATTTCTATGAACTCTGATTCATCTTTTCCTGGAATTTCTATAATGGAAAGTCCATTAGAAAGTCAGCCCTTAGATTCAGATAGAAGCATCAAAGAATCCTCTTTTGAAGAATCAAATATTGAAGATCCACTTATTGTAACACCAGATTGCCAAGAAAAGACCTCACCAAAAGGTGTCGAGAACCCTGCTGTACAAGAGAGTAACCAAAAAATGTTAGGTCCTCCTTTGGAGGTGCTGAAAACGTTAGCCTCTAAAAGAAATGCTGTTGCTTTTCGAAGTTTTAACAGTCATATTAATGCATCCAATAACTCAGAACCATCCAGAATGAACATGACTTCTTTAGATGCAATGGATATTTCGTGTGCCTACAGTGGTTCATATCCCATGGCTATAACCCCTACTCAAAAAAGAAGATCCTGTATGCCACATCAGCAGACCCCAAATCAGATCAAGTCGGGAACTCCATACCGAACTCCGAAGAGTGTGAGAAGAGGGGTGGCCCCCGTTGATGATGGGCGAATTCTAGGAACCCCAGACTACCTTGCACCTGAGCTGTTACTAGGCAGGGCCCATGGTCCTGCGGTAGACTGGTGGGCACTTGGAGTTTGCTTGTTTGAATTTCTAACAGGAATTCCCCCTTTCAATGATGAAACACCACAACAAGTATTCCAGAATATTCTGAAAAGAGATATCCCTTGGCCAGAAGGTGAAGAAAAGTTATCTGATAATGCTCAAAGTGCAGTAGAAATACTTTTAACCATTGATGATACAAAGAGAGCTGGAATGAAAGAGCTAAAACGTCATCCTCTCTTCAGTGATGTGGACTGGGAAAATCTGCAGCATCAGACTATGCCTTTCATCCCCCAGCCAGATGATGAAACAGATACCTCCTATTTTGAAGCCAGGAATACTGCTCAGCACCTGACTGTATCTGGATTTAGTCTGTAGCACAAAAATTTTCCTTTTAGTCTAGCCTTGTGTTATAGAATGAACTTGCATAATTATATACTCCTTAATACTAGATTGATCTAAGGGGGAAAGATCATTATTTAACCTAGTTCAATGTGCTTTTAATGTACGTTACAGCTTTCACAGAGTTAAAAGGCTGAAAGGAATATAGTCAGTAATTTATCTTAACCTCAAAACTGTATATAAATCTTCAAAGCTTTTTTCATTTATTTATTTTGTTTATTGCACTTTATGAAAACTGAAGCATCAATAAAATTAGAGGACACTATTGAGAGTGAGCCACTAGCTTGATTTTCTTTCTCCTCTGATTTCAGTTCACTGTTCAGTTTAGCATTAAAATAATAAAATAATCATACAGTTCCATTTGATTGTAATTTACTTGCATATGATTACCTTCAAAAGCTGTGTTCTTGAGGGCAATCATTTAAGGCCATACTTGTGACCTCAGAAATGCAGCTGGAAGGCCGGGCGCAGTGGCTCACGCATGTAATCCCAGCACTGTGGGAGGCCGAGGCAGGCTGATCTTGAGGTCAGGAGTTCAAGACCATCCTGGCCAACATAATGAAACCCCATCTCTACTGTAAATACACAAATTAGCCGGGCATGTTGGTGGGCGCCTGTAGTCCCAGCTACTCTGGAGGCTGAGGCAGGAGAATCGCTGGAACCCAGGAGGAGGAGGTTGTGGTGAGCCAAGATCGTGCCACTGCACTCCAGCCTGGGCAACAGAACAAGACTCCGTCTCAAAAAAATAAAGGCAGCTGGAAAATTCCTTCACTTAGTAACTCCTTTGGGGAATAAGGAATAAACAAAAAGACTAAGGCTTTCACTTACTTTACTAATGATACGGTTTTGCCACCTTAATTTCTATTGGTGCGTAGTCCACAATATGTATTTGTTTAAAATGGTACTGGAGGACGGGTGCGATGGCTCATGCCTGTAATCCCGGCACTTTGGGAGGCTGAGGCAGGCAAGTCGCTTGAGGCCTGGAGTTTGAGACCAGCCTGGCCAACATGGCAAAACCTGGTCTCTACTAAAAATAGAAAAATTAGCCAGGTATGGTGACGCACACCTGTAGTCCCAGCTACTTGGGAGGCTGAGGCATGAGAATCGCTTGAACCCAGGAGGTGGAGATTGCAGTGAGCTGAGATCACATCACTGCATTTCAGCCTGGGCAACAGAATGAGACTCAGTCTCAAAAAAAAAAAAAATACTACTGGAACAATTAGGTGTGATATAATGTAGCAATTTGAATAGCTTGTTCAATTTTAATAGAAAATGTGATGCTTTACATCTTATTGTCATTTATTTAAAACATGCAACATTATCAGAAACAGATTGGTCCTTAAAAAACAAAACATAATTTTCTTTTGCTTAATAAATATAACTATCACAAGTA\n"
     ]
    }
   ],
   "source": [
    "# Step1 Reading the file:\n",
    "\n",
    "fasta_dict = {}\n",
    "with open('MASTL.fasta') as file:\n",
    "\n",
    "    lines = file.read().split('\\n')\n",
    "    for line in lines:\n",
    "        if line.startswith('>'):\n",
    "            fasta_dict['metadata'] = line\n",
    "            fasta_dict['sequence'] = ''\n",
    "        else:\n",
    "            fasta_dict['sequence'] += line\n",
    "    fasta_dict['sequence'] = fasta_dict['sequence'].replace('\\n', '')\n",
    "print(fasta_dict['metadata'])\n",
    "print(fasta_dict['sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGATCCCACCGCGGGAAGCAAGAAGGAGCCTGGAGGAGGCGCGGCGACTGAGGAGGGCGTGAATAGGATCGCAGTGCCAAAACCGCCCTCCATTGAGGAATTCAGCATAGTGAAGCCCATTAGCCGGGGCGCCTTCGGGAAAGTGTATCTGGGGCAGAAAGGCGGCAAATTGTATGCAGTAAAGGTTGTTAAAAAAGCAGACATGATCAACAAAAATATGACTCATCAGGTCCAAGCTGAGAGAGATGCACTGGCACTAAGCAAAAGCCCATTCATTGTCCATTTGTATTATTCACTGCAGTCTGCAAACAATGTCTACTTGGTAATGGAATATCTTATTGGGGGAGATGTCAAGTCTCTCCTACATATATATGGTTATTTTGATGAAGAGATGGCTGTGAAATATATTTCTGAAGTAGCACTGGCTCTAGACTACCTTCACAGACATGGAATCATCCACAGGGACTTGAAACCGGACAATATGCTTATTTCTAATGAGGGTCATATTAAACTGACGGATTTTGGCCTTTCAAAAGTTACTTTGAATAGAGATATTAATATGATGGATATCCTTACAACACCATCAATGGCAAAACCTAGACAAGATTATTCAAGAACCCCAGGACAAGTGTTATCGCTTATCAGCTCGTTGGGATTTAACACACCAATTGCAGAAAAAAATCAAGACCCTGCAAACATCCTTTCAGCCTGTCTGTCTGAAACATCACAGCTTTCTCAAGGACTCGTATGCCCTATGTCTGTAGATCAAAAGGACACTACGCCTTATTCTAGCAAATTACTAAAATCATGTCTTGAAACAGTTGCCTCCAACCCAGGAATGCCTGTGAAGTGTCTAACTTCTAATTTACTCCAGTCTAGGAAAAGGCTGGCCACATCCAGTGCCAGTAGTCAATCCCACACCTTCATATCCAGTGTGGAATCAGAATGCCACAGCAGTCCCAAATGGGAAAAAGATTGCCAGGAAAGTGATGAAGCATTGGGCCCAACAATGATGAGTTGGAATGCAGTTGAAAAGTTATGCGCAAAATCTGCAAATGCCATTGAGACGAAAGGTTTCAATAAAAAGGATCTGGAGTTAGCTCTTTCTCCCATTCATAACAGCAGTGCCCTTCCCACCACTGGACGCTCTTGTGTAAACCTTGCTAAAAAATGCTTCTCTGGGGAAGTTTCTTGGGAAGCAGTAGAACTGGATGTAAATAATATAAATATGGACACTGACACAAGTCAGTTAGGTTTCCATCAGTCAAATCAGTGGGCTGTGGATTCTGGTGGGATATCTGAAGAGCACCTTGGGAAAAGAAGTTTAAAAAGAAATTTTGAGTTGGTTGACTCCAGTCCTTGTAAAAAAATTATACAGAATAAAAAAACTTGTGTAGAGTATAAGCATAACGAAATGACAAATTGTTATACAAATCAAAATACAGGCTTAACAGTTGAAGTGCAGGACCTTAAGCTATCAGTGCACAAAAGTCAACAAAATGACTGTGCTAATAAGGAGAACATTGTCAATTCTTTTACTGATAAACAACAAACACCAGAAAAATTACCTATACCAATGATAGCAAAAAACCTTATGTGTGAACTCGATGAAGACTGTGAAAAGAATAGTAAGAGGGACTACTTAAGTTCTAGTTTTCTATGTTCTGATGATGATAGAGCTTCTAAAAATATTTCTATGAACTCTGATTCATCTTTTCCTGGAATTTCTATAATGGAAAGTCCATTAGAAAGTCAGCCCTTAGATTCAGATAGAAGCATCAAAGAATCCTCTTTTGAAGAATCAAATATTGAAGATCCACTTATTGTAACACCAGATTGCCAAGAAAAGACCTCACCAAAAGGTGTCGAGAACCCTGCTGTACAAGAGAGTAACCAAAAAATGTTAGGTCCTCCTTTGGAGGTGCTGAAAACGTTAGCCTCTAAAAGAAATGCTGTTGCTTTTCGAAGTTTTAACAGTCATATTAATGCATCCAATAACTCAGAACCATCCAGAATGAACATGACTTCTTTAGATGCAATGGATATTTCGTGTGCCTACAGTGGTTCATATCCCATGGCTATAACCCCTACTCAAAAAAGAAGATCCTGTATGCCACATCAGCAGACCCCAAATCAGATCAAGTCGGGAACTCCATACCGAACTCCGAAGAGTGTGAGAAGAGGGGTGGCCCCCGTTGATGATGGGCGAATTCTAGGAACCCCAGACTACCTTGCACCTGAGCTGTTACTAGGCAGGGCCCATGGTCCTGCGGTAGACTGGTGGGCACTTGGAGTTTGCTTGTTTGAATTTCTAACAGGAATTCCCCCTTTCAATGATGAAACACCACAACAAGTATTCCAGAATATTCTGAAAAGAGATATCCCTTGGCCAGAAGGTGAAGAAAAGTTATCTGATAATGCTCAAAGTGCAGTAGAAATACTTTTAACCATTGATGATACAAAGAGAGCTGGAATGAAAGAGCTAAAACGTCATCCTCTCTTCAGTGATGTGGACTGGGAAAATCTGCAGCATCAGACTATGCCTTTCATCCCCCAGCCAGATGATGAAACAGATACCTCCTATTTTGAAGCCAGGAATACTGCTCAGCACCTGACTGTATCTGGATTTAGTCTGTAG\n"
     ]
    }
   ],
   "source": [
    "dna_sequence = fasta_dict['sequence']\n",
    "\n",
    "# Define the stop codons\n",
    "stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "\n",
    "# Initialize list to hold all open reading frames (ORFs)\n",
    "orfs = []\n",
    "\n",
    "# Find all ATG indices (start codons)\n",
    "start_indices = []\n",
    "index = dna_sequence.find('ATG')\n",
    "while index != -1:\n",
    "    start_indices.append(index)\n",
    "    index = dna_sequence.find('ATG', index + 1)\n",
    "\n",
    "# For each start codon, find the first stop codon in frame\n",
    "for start_index in start_indices:\n",
    "    for i in range(start_index, len(dna_sequence), 3):\n",
    "        codon = dna_sequence[i:i+3]\n",
    "        if codon in stop_codons:\n",
    "            orfs.append(dna_sequence[start_index:i+3])\n",
    "            break\n",
    "\n",
    "# Find the longest ORF\n",
    "longest_orf = max(orfs, key=len) if orfs else ''\n",
    "print(longest_orf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NM_001172303.3 Homo sapiens microtubule associated serine/threonine kinase like (MASTL), transcript variant 1, mRNA\n",
      "ATGGATCCCACCGCGGGAAGCAAGAAGGAGCCTGGAGGAGGCGCGGCGACTGAGGAGGGCGTGAATAGGATCGCAGTGCCAAAACCGCCCTCCATTGAGGAATTCAGCATAGTGAAGCCCATTAGCCGGGGCGCCTTCGGGAAAGTGTATCTGGGGCAGAAAGGCGGCAAATTGTATGCAGTAAAGGTTGTTAAAAAAGCAGACATGATCAACAAAAATATGACTCATCAGGTCCAAGCTGAGAGAGATGCACTGGCACTAAGCAAAAGCCCATTCATTGTCCATTTGTATTATTCACTGCAGTCTGCAAACAATGTCTACTTGGTAATGGAATATCTTATTGGGGGAGATGTCAAGTCTCTCCTACATATATATGGTTATTTTGATGAAGAGATGGCTGTGAAATATATTTCTGAAGTAGCACTGGCTCTAGACTACCTTCACAGACATGGAATCATCCACAGGGACTTGAAACCGGACAATATGCTTATTTCTAATGAGGGTCATATTAAACTGACGGATTTTGGCCTTTCAAAAGTTACTTTGAATAGAGATATTAATATGATGGATATCCTTACAACACCATCAATGGCAAAACCTAGACAAGATTATTCAAGAACCCCAGGACAAGTGTTATCGCTTATCAGCTCGTTGGGATTTAACACACCAATTGCAGAAAAAAATCAAGACCCTGCAAACATCCTTTCAGCCTGTCTGTCTGAAACATCACAGCTTTCTCAAGGACTCGTATGCCCTATGTCTGTAGATCAAAAGGACACTACGCCTTATTCTAGCAAATTACTAAAATCATGTCTTGAAACAGTTGCCTCCAACCCAGGAATGCCTGTGAAGTGTCTAACTTCTAATTTACTCCAGTCTAGGAAAAGGCTGGCCACATCCAGTGCCAGTAGTCAATCCCACACCTTCATATCCAGTGTGGAATCAGAATGCCACAGCAGTCCCAAATGGGAAAAAGATTGCCAGGAAAGTGATGAAGCATTGGGCCCAACAATGATGAGTTGGAATGCAGTTGAAAAGTTATGCGCAAAATCTGCAAATGCCATTGAGACGAAAGGTTTCAATAAAAAGGATCTGGAGTTAGCTCTTTCTCCCATTCATAACAGCAGTGCCCTTCCCACCACTGGACGCTCTTGTGTAAACCTTGCTAAAAAATGCTTCTCTGGGGAAGTTTCTTGGGAAGCAGTAGAACTGGATGTAAATAATATAAATATGGACACTGACACAAGTCAGTTAGGTTTCCATCAGTCAAATCAGTGGGCTGTGGATTCTGGTGGGATATCTGAAGAGCACCTTGGGAAAAGAAGTTTAAAAAGAAATTTTGAGTTGGTTGACTCCAGTCCTTGTAAAAAAATTATACAGAATAAAAAAACTTGTGTAGAGTATAAGCATAACGAAATGACAAATTGTTATACAAATCAAAATACAGGCTTAACAGTTGAAGTGCAGGACCTTAAGCTATCAGTGCACAAAAGTCAACAAAATGACTGTGCTAATAAGGAGAACATTGTCAATTCTTTTACTGATAAACAACAAACACCAGAAAAATTACCTATACCAATGATAGCAAAAAACCTTATGTGTGAACTCGATGAAGACTGTGAAAAGAATAGTAAGAGGGACTACTTAAGTTCTAGTTTTCTATGTTCTGATGATGATAGAGCTTCTAAAAATATTTCTATGAACTCTGATTCATCTTTTCCTGGAATTTCTATAATGGAAAGTCCATTAGAAAGTCAGCCCTTAGATTCAGATAGAAGCATCAAAGAATCCTCTTTTGAAGAATCAAATATTGAAGATCCACTTATTGTAACACCAGATTGCCAAGAAAAGACCTCACCAAAAGGTGTCGAGAACCCTGCTGTACAAGAGAGTAACCAAAAAATGTTAGGTCCTCCTTTGGAGGTGCTGAAAACGTTAGCCTCTAAAAGAAATGCTGTTGCTTTTCGAAGTTTTAACAGTCATATTAATGCATCCAATAACTCAGAACCATCCAGAATGAACATGACTTCTTTAGATGCAATGGATATTTCGTGTGCCTACAGTGGTTCATATCCCATGGCTATAACCCCTACTCAAAAAAGAAGATCCTGTATGCCACATCAGCAGACCCCAAATCAGATCAAGTCGGGAACTCCATACCGAACTCCGAAGAGTGTGAGAAGAGGGGTGGCCCCCGTTGATGATGGGCGAATTCTAGGAACCCCAGACTACCTTGCACCTGAGCTGTTACTAGGCAGGGCCCATGGTCCTGCGGTAGACTGGTGGGCACTTGGAGTTTGCTTGTTTGAATTTCTAACAGGAATTCCCCCTTTCAATGATGAAACACCACAACAAGTATTCCAGAATATTCTGAAAAGAGATATCCCTTGGCCAGAAGGTGAAGAAAAGTTATCTGATAATGCTCAAAGTGCAGTAGAAATACTTTTAACCATTGATGATACAAAGAGAGCTGGAATGAAAGAGCTAAAACGTCATCCTCTCTTCAGTGATGTGGACTGGGAAAATCTGCAGCATCAGACTATGCCTTTCATCCCCCAGCCAGATGATGAAACAGATACCTCCTATTTTGAAGCCAGGAATACTGCTCAGCACCTGACTGTATCTGGATTTAGTCTGTAG\n"
     ]
    }
   ],
   "source": [
    "from dna_sequence import DnaSequence\n",
    "\n",
    "dna_sequence = DnaSequence('MASTL.fasta')\n",
    "print(dna_sequence.metadata)\n",
    "print(dna_sequence.orf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "data-analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
